### PR TITLE
feat: complete the queries

### PR DESCRIPTION
**Pull request by amaanq**

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -1,6 +1,9 @@
 /**
  * @file Pony grammar for tree-sitter
  * @author Amaan Qureshi <amaanq12@gmail.com>
+ * @author Matthias Wahl <matthiaswahl@m7w3.de>
```
Comment (mfelsche) [inline grammar.js:4]: :heart:

File: grammar.js [hunk 2]
```diff
@@ -350,7 +362,7 @@ module.exports = grammar({
       $.parenthesized_expression,
       $.tuple_expression,
       $.generic_expression,
-      $.literal,
+      // $.literal,
```
Comment (mfelsche) [inline grammar.js]: it is actually possible to call functions on literals like strings and array literals.
E.g.:

```pony
actor Main
  new create(env: Env) =>
    [as U8:].size()
```

File: queries/highlights.scm [hunk 3]
```diff
@@ -1,132 +1,271 @@
...
-  "struct"
   "primitive"
   "interface"
   "trait"
-  "type"
+  "struct"
+  "embed"
+  "let"
+  "var"
+  "be"
```
Comment (mfelsche) [inline queries/highlights.scm]: i'd capture this as `keyword.function`. It is defining a behavior, which is in some sense different to a function, but looks very much alike.

File: queries/highlights.scm [hunk 4]
```diff
@@ -1,132 +1,271 @@
...
+(try_statement "end" @exception)
+
+(recover_expression "end" @exception)
+
+; Attributes
+
+(annotation) @attribute
+
+[
   "#read"
```
Comment (mfelsche) [inline queries/highlights.scm]: those are generic capabilities, all defining a certain set of possible capabilities. I'd put them underneath `type.qualifier` like the others (tag, iso, etc)

File: queries/highlights.scm [hunk 5]
```diff
@@ -1,132 +1,271 @@
...
+; Conditionals
+
+[
+  "if"
+  "ifdef"
+  "iftype"
+  "then"
+  "else"
+  "elseif"
+  "until"
```
Comment (mfelsche) [inline queries/highlights.scm:67]: i think this should be in the `repeat` section.
It is only used here:

```pony
repeat
   env.out.write(".")
until enough == true end
```
so it should also be marked as such.

File: queries/highlights.scm [hunk 6]
```diff
@@ -1,132 +1,271 @@
...
+  "where"
+] @keyword
+
+[
+  "fun"
+] @keyword.function
+
+[
+  "new"
+  "in"
```
Comment (mfelsche) [inline queries/highlights.scm:34]: this is only used in `for` loops, maybe then it should also be part of `repeat`?

File: queries/highlights.scm [hunk 7]
```diff
@@ -1,132 +1,271 @@
...
+  "object"
+  "where"
+] @keyword
+
+[
+  "fun"
+] @keyword.function
+
+[
+  "new"
```
Comment (mfelsche) [inline queries/highlights.scm]: this is marking the beginning of a constructor. It is never used as an operator. So I'd argue for `keyword.function` here.

File: queries/highlights.scm [hunk 8]
```diff
@@ -1,132 +1,271 @@
...
+[
+  "if"
+  "ifdef"
+  "iftype"
+  "then"
+  "else"
+  "elseif"
+  "until"
+  "match"
+  "break"
```
Comment (mfelsche) [inline queries/highlights.scm]: `break` is only used in loops, so maybe it should be put underneath`repeat`?

File: queries/highlights.scm [hunk 9]
```diff
@@ -1,132 +1,271 @@
...
+
+(do_block "end" @repeat)
+
+(repeat_statement "end" @repeat)
+
+; Exceptions
+
+[
+  "try"
+  "recover"
```
Comment (mfelsche) [inline queries/highlights.scm]: `recover` is not used for recovering from errors, but to transform references to have another capability.
E.g. we can use it to turn a freshly created string from a `ref^` into an `iso^` capability:

```pony
let iso_str: String iso =
  recover iso
     String.create(5) .> append("hello")
  end
```

so it should maybe go underneath `keyword`.

File: queries/highlights.scm [hunk 10]
```diff
@@ -1,132 +1,271 @@
...
+] @keyword
+
+[
+  "fun"
+] @keyword.function
+
+[
+  "new"
+  "in"
+  "is"
```
Comment (mfelsche) [inline queries/highlights.scm:35]: Some missing keywords:

```suggestion
  "is"
  "compile_intrinsic"
```

File: queries/highlights.scm [hunk 11]
```diff
@@ -1,132 +1,271 @@
...
+; Exceptions
+
+[
+  "try"
+  "recover"
+  (error)
+  "compile_error"
+] @exception
+
+(try_statement "end" @exception)
```
Comment (mfelsche) [inline queries/highlights.scm:100]: try blocks also have `else` and `then`:

```pony
try
  i_might_error()?
else
  "oh noes, it did error!"
then
  // like `finally` in java
  "anyways"
end
```

File: queries/highlights.scm [hunk 12]
```diff
@@ -1,132 +1,271 @@
...
+
+(member_expression "." (identifier) @field)
+
+; Constructors
+
+(constructor (identifier) @constructor)
+
+; Methods
+
+(method (identifier) @method)
```
Comment (mfelsche) [inline queries/highlights.scm:126]: For highlighting purposes, i'd treat `behaviors` the same as `methods`:

```suggestion
(method (identifier) @method)
(behavior (identifier) @method)
```

File: queries/highlights.scm [hunk 13]
```diff
@@ -1,132 +1,271 @@
...
+
+; Repeats
+
+[
+  "repeat"
+  "while"
+  "for"
+  "continue"
+  "do"
+] @repeat
```
Comment (mfelsche) [inline queries/highlights.scm:86]: we totally forgot, that all the loop constructs can have an `else` clause for the case when there is no iteration done. Will add this as another PR.

**Review by mfelsche — CHANGES_REQUESTED** [review]
I left some comments, as I would move some stuff to different categories.

File: queries/highlights.scm (hunk 10 above)
Comment (mfelsche) [reply queries/highlights.scm:35]: Ignore this, I have just seen that it is added above.

File: queries/highlights.scm (hunk 3 above)
Comment (amaanq) [reply queries/highlights.scm]: Do you mean just behavior? If so yes

File: queries/highlights.scm (hunk 5 above)
Comment (amaanq) [reply queries/highlights.scm:67]: until precedes a condition though, fitting `@conditional` more imo

File: queries/highlights.scm (hunk 6 above)
Comment (amaanq) [reply queries/highlights.scm:34]: no, it's typical to have such keywords like 'in' (similar to python's `for x in y`) as `@keyword.operator`

File: queries/highlights.scm (hunk 7 above)
Comment (amaanq) [reply queries/highlights.scm]: I applied similar JavaScript convention (`const a = new A()`) where A is highlighted as constructor, but new `@keyword.operator`, so I think it makes sense again here

File: queries/highlights.scm (hunk 8 above)
Comment (amaanq) [reply queries/highlights.scm]: Sure, it's not used in match though?

File: queries/highlights.scm (hunk 7 above)
Comment (mfelsche) [reply queries/highlights.scm]: In javascript you create objects with `new`, so there it makes sense as an operator. In pony, `new` is only a keyword to mark the start of a constructor definition. So it is not an operator at all. Example:

```pony
class X
  new constructor() => None // constructor definition
  
X.constructor() // instance creation, calling the constructor
```

File: queries/highlights.scm (hunk 6 above)
Comment (mfelsche) [reply queries/highlights.scm:34]: If this is the case then lets go with this.

File: queries/highlights.scm (hunk 5 above)
Comment (mfelsche) [reply queries/highlights.scm:67]: that is a totally sensible way of thinking about it.

File: queries/highlights.scm (hunk 8 above)
Comment (mfelsche) [reply queries/highlights.scm]: No, match does not fall through, so doesn't need a `break`. It is only used in loops.

File: queries/highlights.scm (hunk 3 above)
Comment (mfelsche) [reply queries/highlights.scm]: yes, only `be`

**Comment by mfelsche** [comment]
The scanner.c file needs to be added to the binding.gyp for the windows tests to work correctly.

File: queries/highlights.scm (hunk 11 above)
Comment (amaanq) [reply queries/highlights.scm:100]: hmm good point

File: queries/highlights.scm (hunk 7 above)
Comment (amaanq) [reply queries/highlights.scm]: Hmm, are new and constructor **always** used together? It feels weird to have both be the same from a non-Pony user's pov

I made it `@constructor` now anyways

File: queries/highlights.scm (hunk 13 above)
Comment (amaanq) [reply queries/highlights.scm:86]: Ahh..is it after the do block?

Comment (mfelsche) [reply queries/highlights.scm:86]: yeah, exactly like `else` blocks in all the `if` variants.

**Review by mfelsche — APPROVED** [review]
LGTM

**Pull request merged**